src/hooks/useMessagesLane.ts

```diff
@@ -33,77 +33,84 @@ interface Props {
-            const latest_generated_nonce = res.get('latest_generated_nonce').toString();
-            const latest_received_nonce = res.get('latest_received_nonce').toString();
-            const pendingMessages = new BN(latest_generated_nonce).sub(new BN(latest_received_nonce));
+  const { bridgedMessages } = getSubstrateDynamicNames(chain);
 
-            setter({
-              latestReceivedNonce: latest_received_nonce.toString(),
-              pendingMessages: pendingMessages.isNeg() ? '0' : pendingMessages.toString(),
-              totalMessages: latest_generated_nonce
-            });
+  useEffect(() => {
+    let unsubscribeOutboundLanes: () => void;
+    let unsubscribeInboundLanes: () => void;
+
+    if (api && isApiReady && api.query[bridgedMessages] && chain && isMounted()) {
+      api.query[bridgedMessages]
+        .outboundLanes(laneId, (res: any) => {
+          const latest_generated_nonce = res.get('latest_generated_nonce').toString();
+          const latest_received_nonce = res.get('latest_received_nonce').toString();
+          const pendingMessages = new BN(latest_generated_nonce).sub(new BN(latest_received_nonce));
+
+          setOutboudLanes({
+            latestReceivedNonce: latest_received_nonce.toString(),
+            pendingMessages: pendingMessages.isNeg() ? '0' : pendingMessages.toString(),
+            totalMessages: latest_generated_nonce
           });
-          return Promise.resolve(u);
-        }
-        const u = await api.query[bridgedMessages].inboundLanes(laneId, (res: any) => {
-          setter({ bridgeReceivedMessages: res.get('last_confirmed_nonce').toString() });
+        })
+        .then((unsub) => {
+          unsubscribeOutboundLanes = unsub;
+        })
+        .catch((message) => {
+          dispatchMessage(MessageActionsCreators.triggerErrorMessage({ message }));
+          logger.error(message);
         });
-        return Promise.resolve(u);
-      } catch (e) {
-        dispatchMessage(MessageActionsCreators.triggerErrorMessage({ message: e.message }));
-        logger.error(e.message);
-        return Promise.reject();
-      }
-    };
 
-    const shouldGetLane: boolean = !!(api && isApiReady && api.query[bridgedMessages] && chain);
-    const unsubscribeOutboundLanes: Promise<VoidFn> | null = shouldGetLane ? getLane(setOutboudLanes, true) : null;
-    const unsubscribeInboundLanes: Promise<VoidFn> | null = shouldGetLane ? getLane(setInboudLanes) : null;
+      api.query[bridgedMessages]
+        .inboundLanes(laneId, (res: any) => {
+          setBridgesReceivedMessages(res.get('last_confirmed_nonce').toString());
+        })
+        .then((unsub) => {
+          unsubscribeInboundLanes = unsub;
+        })
+        .catch((message) => {
+          dispatchMessage(MessageActionsCreators.triggerErrorMessage({ message }));
+          logger.error(message);
+        });
+    }
 
-    return async (): Promise<void> => {
```

Comment:
This is not right , the return function should not be awaiting to resolve an async function. So i removed the async and call it as it was before.

Reply:
No, I'm pretty sure you are incorrect.
```js
return async () => {
  unsubscribeOutboundLanes && (await unsubscribeOutboundLanes)();
  unsubscribeInboundLanes && (await unsubscribeInboundLanes)();
};
```
is (almost) exactly the same as:
```js
return () => {
   unsubscribeOutboundLanes && unsubscribeOutboundLanes.then(u => u());
   unsubscribeInboundLanes && unsubscribeInboundLanes.then(u => u());
};
```
The only differences are the return value (in the first case it's going to be a `Promise<void>` in the second case it's going to be `void`) and the unsubscribe ordering (await is going to be sequential, the other parallel) essentialy it's going to do exactly the same things.

I'd vote for reverting all of the code changes here, cause the new code is imho much less readable and definitely less concise.

Reply:
You are missing the point here. Of course `async/await` in `useEffect` is causing us troubles - that's the exact reason why we have this discussion, we are thinking about abstracting this out and we have leaks due to subscriptions or setState-after-unmount. But using `then/catch` instead of `async/await` is not solving anything - it's just different syntax for the same stuff.

With current approach we **must** do asynchronous clean up and my point is that there is no difference between having a function and chaining promises or writing `async/await` function - the effect is exactly the same we wait for the promise to resolve and then we unsubscribe.

Reply:
It is not exactly the same because of the type of the function that you are returning. With async await, you are returning a promise type instead of void. And i think that's not right ... please check the `EffectCallback` type of react code and read the comment:

```
    // NOTE: callbacks are _only_ allowed to return either void, or a destructor.
    // The destructor is itself only allowed to return void.
    type EffectCallback = () => (void | (() => void | undefined));
```

With `async/await` the function is returning a Promise type and the comment clearly say that must return void.

Reply:
Okay, I'm super confused now. We do use `TypeScript`, right? If the compiler does not complaint about incorrect types then something is really wrong here.

Yes, if the function suggests the return type should be `void` - that's fine. The only difference is that you need to introduce an `async` block if that makes the code simpler and more readable compared to nested `then-catch`.
Please find a helper function here to make sure the types are correct:
```js
function makeVoid(fn: () => Promise<void>): () => void {
  return () => {
     fn().catch(e => logger.error('e'));
  };
}
```

You can use this function everywhere like that:
```js
useEffect(makeVoid(async () => {
  /// whatever
}));
```
and and awesome consequence is that you don't even need `try-catch` block, cause it logs the error and if you don't see the code becoming cleaner if we do this, then the rest of the discussion is not worth it and we can just stay with `then-catch` if you prefer.

Reply:
i like what you propose , but i had to adjust makeVoid a little to:

```
function makeVoid(fn: () => Promise<(() => void) | undefined>): () => void {
  return () => {
    fn().catch((e) => logger.error(e));
  };
}
```
In any case i receive the following eslint warning:

```
Accepts a function that contains imperative, possibly effectful code.

@param effect — Imperative function that can return a cleanup function

@param deps — If present, effect will only activate if the values in the list change.

@version — 16.8.0

@see — https ://reactjs.org/docs/hooks-reference.html#useeffect

React Hook useEffect received a function whose dependencies are unknown. Pass an inline function instead.eslintreact-hooks/exhaustive-deps
```

Reply:
i like what you propose , but i had to adjust makeVoid a little to:

```
function makeVoid(fn: () => Promise<(() => void) | undefined>): () => void {
  return () => {
    fn().catch((e) => logger.error(e));
  };
}
```

In any case i receive the following eslint warning:

```
Accepts a function that contains imperative, possibly effectful code.

@param effect — Imperative function that can return a cleanup function

@param deps — If present, effect will only activate if the values in the list change.

@version — 16.8.0

@see — https ://reactjs.org/docs/hooks-reference.html#useeffect

React Hook useEffect received a function whose dependencies are unknown. Pass an inline function instead.eslintreact-hooks/exhaustive-deps
```

Reply:
Please check this proposal: https://github.com/paritytech/parity-bridges-ui/pull/181